File: ghost/client-core/net/tests/entitlement_crypto.rs
//! The stateless `EntitlementCrypto` functions (design §11.7) over the test schedule: the verified
//! summary, layout digests, offline token checks, address validation and the payment URI; and the
//! production load path, which never accepts a test schedule.

mod common;

use common::*;
use ed25519_dalek::Signer as _;
use ghost_client_net::entitlement::{self, load_schedule, Product, EMBEDDED_SCHEDULE};
use ghost_entitlement::batch::Layout;
use ghost_entitlement::monero::{AddressError, AddressPurpose};
use ghost_entitlement::{Kind, ScheduleError};

#[test]
fn the_production_load_path_refuses_every_test_schedule() {
    assert_eq!(
        load_schedule(TEST_SCHEDULE).err(),
        Some(ScheduleError::NoPinnedKey)
    );
    // The test content under the stagenet network byte, signed with the test key: the pinned
    // stagenet key refuses it.
    let relabelled = resigned(|c| c.network = ghost_entitlement::monero::MoneroNetwork::Stagenet);
    let content = relabelled.content();
    let sig = schedule_signing_key().sign(&content.signing_message().unwrap());
    let bytes = content.to_signed_bytes(&sig.to_bytes()).unwrap();
    assert_eq!(load_schedule(&bytes).err(), Some(ScheduleError::Signature));
    // The embedded schedule is a different file from the test schedule.
    assert_ne!(EMBEDDED_SCHEDULE, TEST_SCHEDULE);
}

/// A strict reader of the summary layout (the rules of the Kotlin decoder).
struct Summary {
    slots: Vec<(u8, u64, u64, String)>,
    prices: Vec<(u64, u64)>,
    keys: Vec<(u8, u64, [u8; 32])>,
    revoked: Vec<(u8, u64)>,
}

fn read(b: &[u8]) -> Summary {
    let u64_at = |i: usize| u64::from_be_bytes(b[i..i + 8].try_into().unwrap());
    let u16_at = |i: usize| usize::from(u16::from_be_bytes([b[i], b[i + 1]]));
    let count = usize::from(b[77]);
    let mut i = 78;
    let mut slots = Vec::new();
    for _ in 0..count {
        let len = usize::from(b[i + 17]);
        let onion = String::from_utf8(b[i + 18..i + 18 + len].to_vec()).unwrap();
        slots.push((b[i], u64_at(i + 1), u64_at(i + 9), onion));
        i += 18 + len;
    }
    let n = u16_at(i);
    i += 2;
    let prices = (0..n)
        .map(|k| (u64_at(i + 16 * k), u64_at(i + 16 * k + 8)))
        .collect();
    i += 16 * n;
    let n = u16_at(i);
    i += 2;
    let keys = (0..n)
        .map(|k| {
            let at = i + 41 * k;
            (
                b[at],
                u64_at(at + 1),
                b[at + 9..at + 41].try_into().unwrap(),
            )
        })
        .collect();
    i += 41 * n;
    let n = u16_at(i);
    i += 2;
    let revoked = (0..n)
        .map(|k| (b[i + 9 * k], u64_at(i + 9 * k + 1)))
        .collect();
    i += 9 * n;
    assert_eq!(i, b.len(), "no trailing bytes");
    Summary {
        slots,
        prices,
        keys,
        revoked,
    }
}

#[test]
fn the_summary_carries_every_fact_kotlin_remembers() {
    let s = schedule();
    let b = entitlement::schedule_summary(s).unwrap();
    assert_eq!(&b[..32], s.digest());
    assert_eq!(u64::from_be_bytes(b[32..40].try_into().unwrap()), 1);
    assert_eq!(b[40], 3, "regtest");
    assert_eq!(u64::from_be_bytes(b[41..49].try_into().unwrap()), 2957);
    assert_eq!(u64::from_be_bytes(b[49..57].try_into().unwrap()), 2982);
    assert_eq!(&b[57..62], &[10, 0x02, 0xD0, 0x08, 0x70]); // 10 confirmations, 720, 2160
    assert_eq!(&b[62..69], &[16, 8, 2, 10, 10, 50, 24]);
    assert_eq!(
        u64::from_be_bytes(b[69..77].try_into().unwrap()),
        268_435_456
    );
    let sum = read(&b);
    assert_eq!(sum.slots.len(), 4);
    for (slot, from, until, onion) in &sum.slots {
        let entry = s
            .content()
            .slots
            .iter()
            .find(|e| e.slot == *slot && e.valid_from_week == *from)
            .unwrap();
        assert_eq!(
            (entry.valid_until_week, entry.onion.as_str()),
            (*until, onion.as_str())
        );
        ghost_client_net::OnionAddress::parse(onion).unwrap();
    }
    assert_eq!(
        sum.slots[3],
        (2, 2967, 0, s.slot_onion(2, 2967).unwrap().to_owned())
    );
    assert_eq!(sum.prices.len(), 3);
    assert_eq!(sum.prices[0], (227, 200_000_000_000));
    assert_eq!(sum.keys.len(), s.keys().count());
    assert_eq!(sum.keys.len(), 26 + 7 + 3);
    for (kind, epoch, id) in &sum.keys {
        let k = s.key(Kind::from_byte(*kind).unwrap(), *epoch).unwrap();
        assert_eq!(&k.key_id, id);
    }
    assert!(sum.revoked.is_empty());

    // Revocations travel in ES order.
    let revoked = resigned(|c| c.revoked = vec![(Kind::Access, 2960), (Kind::Credit, 228)]);
    let sum = read(&entitlement::schedule_summary(&revoked).unwrap());
    assert_eq!(sum.revoked, vec![(1, 2960), (3, 228)]);
}

#[test]
fn layout_digests_are_those_of_the_batch_layouts() {
    let s = schedule();
    let n = |b: &[u8; 36]| u32::from_be_bytes(b[32..].try_into().unwrap());
    let cases = [
        (
            Product::PackXmr,
            2957,
            Layout::pack(s, 2957, true).unwrap(),
            243,
        ),
        (
            Product::PackCredits,
            2957,
            Layout::pack(s, 2957, false).unwrap(),
            242,
        ),
        // Slot 2 changes relay at 2967 inside this pack; the layout counts slots, not relays.
        (
            Product::PackXmr,
            2965,
            Layout::pack(s, 2965, true).unwrap(),
            243,
        ),
        (Product::Trial, 2957, Layout::trial(s, 2957).unwrap(), 48),
        (Product::Refresh, 227, Layout::refresh(s, 227).unwrap(), 1),
        (
            Product::PackXmr,
            2978,
            Layout::pack(s, 2978, true).unwrap(),
            243,
        ),
    ];
    for (product, index, layout, count) in cases {
        let b = entitlement::layout_digest(s, product, index).unwrap();
        assert_eq!(b[..32], layout.digest(), "{product:?} {index}");
        assert_eq!(n(&b), count, "{product:?} {index}");
    }
    // Weeks and epochs the ES does not cover have no layout.
    assert!(entitlement::layout_digest(s, Product::PackXmr, 2979).is_err());
    assert!(entitlement::layout_digest(s, Product::Trial, 2982).is_err());
    assert!(entitlement::layout_digest(s, Product::Refresh, 230).is_err());
    assert!(entitlement::layout_digest(s, Product::PackXmr, 2956).is_err());
}

#[test]
fn tokens_are_checked_offline_for_their_kind() {
    let s = schedule();
    let access = mint(Kind::Access, 2959, Some(1), seed(1, 0x7A));
    let invite = mint(Kind::Invite, INVITE_EPOCH, None, seed(2, 0x7A));
    let credit = mint(Kind::Credit, CREDIT_EPOCH, None, seed(3, 0x7A));
    for (token, kind, epoch, slot) in [
        (&access, Kind::Access, 2959, Some(1)),
        (&invite, Kind::Invite, INVITE_EPOCH, None),
        (&credit, Kind::Credit, CREDIT_EPOCH, None),
    ] {
        let v = entitlement::verify_token(s, token.as_bytes(), kind).unwrap();
        assert_eq!(
            (v.kind, v.epoch, v.slot, v.nullifier),
            (kind, epoch, slot, token.nullifier())
        );
        let p = entitlement::pack_verified(&v);
        assert_eq!(p[0], kind.byte());
        assert_eq!(&p[1..9], &epoch.to_be_bytes());
        assert_eq!(p[9], slot.unwrap_or(0xFF));
        assert_eq!(&p[10..], &token.nullifier());
        // Any other kind refuses it.
        for other in Kind::ALL.into_iter().filter(|k| *k != kind) {
            assert_eq!(entitlement::verify_token(s, token.as_bytes(), other), None);
        }
    }
    // A pinned vector token, a tampered one and one of a revoked epoch.
    let a1 = pinned("a1");
    assert_eq!(
        entitlement::verify_token(s, &a1, Kind::Access)
            .unwrap()
            .slot,
        Some(1)
    );
    let mut tampered = invite.as_bytes().to_vec();
    tampered[300] ^= 1;
    assert_eq!(entitlement::verify_token(s, &tampered, Kind::Invite), None);
    let revoked = resigned(|c| c.revoked = vec![(Kind::Invite, INVITE_EPOCH)]);
    assert_eq!(
        entitlement::verify_token(&revoked, invite.as_bytes(), Kind::Invite),
        None
    );
    assert_eq!(entitlement::verify_token(s, &a1[..300], Kind::Access), None);
}

#[test]
fn addresses_of_the_schedule_network_only() {
    let s = schedule();
    assert_eq!(
        entitlement::validate_address(s, SUBADDRESS, AddressPurpose::Invoice),
        Ok(0x0302)
    );
    assert_eq!(
        entitlement::validate_address(s, SUBADDRESS, AddressPurpose::Payout),
        Ok(0x0302)
    );
    assert_eq!(
        entitlement::validate_address(s, PAYOUT, AddressPurpose::Payout),
        Ok(0x0301)
    );
    assert_eq!(
        entitlement::validate_address(s, PAYOUT, AddressPurpose::Invoice),
        Err(AddressError::WrongType)
    );
    assert_eq!(
        entitlement::validate_address(s, STAGENET_SUBADDRESS, AddressPurpose::Payout),
        Err(AddressError::WrongNetwork)
    );
    assert_eq!(
        entitlement::payment_uri(s, SUBADDRESS, 200_000_000_000).unwrap(),
        format!("monero:{SUBADDRESS}?tx_amount=0.200000000000")
    );
    assert!(entitlement::payment_uri(s, PAYOUT, 1).is_err());
    assert!(entitlement::payment_uri(s, SUBADDRESS, 0).is_err());
}

File: ghost/android/network/src/main/kotlin/org/ghost/network/EntitlementCrypto.kt
package org.ghost.network

/**
 * Stateless entitlement functions of the Rust core (Phase 8 design §11.7) over the Entitlement
 * Schedule built into `libghost_client_net.so`, which the native side verifies under the pinned
 * schedule key at first use. Kotlin never handles the ES bytes, keys, blinded messages or blind
 * signatures: it gets the verified [ScheduleSummary], layout digests, offline token verdicts,
 * validated addresses and the payment URI. Native results are fixed-layout byte strings
 * (integers big-endian) decoded strictly here: a length, count or value outside the layout is
 * `malformed_response`. Failures are [NetworkException] categories (`internal` if the embedded
 * schedule does not verify).
 */
object EntitlementCrypto {
    const val PRODUCT_PACK_XMR = 1
    const val PRODUCT_PACK_CREDITS = 2
    const val PRODUCT_TRIAL = 3
    const val PRODUCT_REFRESH = 4

    const val KIND_ACCESS = 1
    const val KIND_INVITE = 2
    const val KIND_CREDIT = 3

    const val PURPOSE_INVOICE = 1
    const val PURPOSE_PAYOUT = 2

    const val ADDRESS_STANDARD = 1
    const val ADDRESS_SUBADDRESS = 2

    /** Slot byte of a token without a relay slot (invite, credit). */
    const val NO_SLOT = 0xFF

    const val TOKEN_BYTES = 354

    private const val MALFORMED = "malformed_response"
    private const val MAX_SLOT = 31

    /** The verified summary of the embedded ES (read it once per process and keep it). */
    fun scheduleSummary(): ScheduleSummary {
        NativeLibrary.ensureLoaded()
        return decodeSummary(nativeScheduleSummary())
    }

    /**
     * The frozen layout of [product] at [index] (the base week of a pack or trial, the credit epoch
     * of a refresh): its digest, stored at purchase time, and its number of positions.
     */
    fun layout(product: Int, index: Long): Layout {
        require(product in PRODUCT_PACK_XMR..PRODUCT_REFRESH) { "unknown product" }
        require(index >= 0) { "index must not be negative" }
        NativeLibrary.ensureLoaded()
        return decodeLayout(nativeLayoutDigest(product, index))
    }

    /**
     * The offline check of a token under the embedded ES for [kind] (access: any slot of its week),
     * or null when the schedule refuses it (unknown or revoked key, another kind, a wrong
     * challenge or authenticator). The acceptance window of the epoch is the caller's check.
     */
    fun verifyToken(token: ByteArray, kind: Int): VerifiedToken? {
        require(kind in KIND_ACCESS..KIND_CREDIT) { "unknown token kind" }
        NativeLibrary.ensureLoaded()
        val raw = try {
            nativeVerifyToken(token, kind)
        } catch (e: NetworkException) {
            if (e.category == "rejected") return null
            throw e
        }
        return decodeVerified(raw, kind)
    }

    /**
     * Validates a Monero address of the ES network for [purpose] (an invoice takes a subaddress, a
     * payout a standard address or a subaddress), or null when it is refused.
     */
    fun validateAddress(address: String, purpose: Int): AddressInfo? {
        require(purpose == PURPOSE_INVOICE || purpose == PURPOSE_PAYOUT) { "unknown address purpose" }
        NativeLibrary.ensureLoaded()
        val code = try {
            nativeValidateAddress(address, purpose)
        } catch (e: NetworkException) {
            if (e.category == "rejected") return null
            throw e
        }
        return decodeAddress(code)
    }

    /** `monero:<subaddress>?tx_amount=<12 decimals>`, built natively from validated values. */
    fun paymentUri(subaddress: String, amountAtomic: Long): String {
        require(amountAtomic > 0) { "amount must be positive" }
        NativeLibrary.ensureLoaded()
        return nativePaymentUri(subaddress, amountAtomic)
    }

    /** Protocol constants of the ES. */
    data class Constants(
        val confirmations: Int,
        val invoiceBlocks: Int,
        val graceBlocks: Int,
        val accessPerSlot: Int,
        val trialPerSlot: Int,
        val invitesPerPack: Int,
        val creditsPerFreePack: Int,
        val minClaimCredits: Int,
        val maxClaimCredits: Int,
        val earlyWindowHours: Int,
        val capabilityQuotaBytes: Long,
    )

    /** Slot [slot] is served by [onion] in weeks `[validFromWeek, validUntilWeek)` (0 = open). */
    data class Slot(val slot: Int, val validFromWeek: Long, val validUntilWeek: Long, val onion: OnionAddress) {
        fun validIn(week: Long): Boolean = validFromWeek <= week && (validUntilWeek == 0L || week < validUntilWeek)
    }

    data class Price(val priceEpoch: Long, val packPriceAtomic: Long)

    /** A key of the ES: its (kind, epoch) and key id (what `ent_key` remembers). */
    class KeyId(val kind: Int, val epoch: Long, keyId: ByteArray) {
        private val id = keyId.copyOf()
        fun keyId(): ByteArray = id.copyOf()
        override fun equals(other: Any?): Boolean =
            other is KeyId && other.kind == kind && other.epoch == epoch && other.id.contentEquals(id)
        override fun hashCode(): Int = 31 * (31 * kind + epoch.hashCode()) + id.contentHashCode()
        override fun toString(): String = "KeyId(kind=$kind, epoch=$epoch)"
    }

    data class Revoked(val kind: Int, val epoch: Long)

    /** The verified ES as Kotlin sees it (rule 5 memory: keys, week slot sets, prices, revocations). */
    class ScheduleSummary(
        digest: ByteArray,
        val seq: Long,
        val network: Int,
        val firstWeek: Long,
        val lastWeek: Long,
        val constants: Constants,
        val slots: List<Slot>,
        val prices: List<Price>,
        val keys: List<KeyId>,
        val revoked: List<Revoked>,
    ) {
        private val digestBytes = digest.copyOf()
        fun digest(): ByteArray = digestBytes.copyOf()

        /** Slot numbers valid in [week], ascending (the layout order, design §4.2). */
        fun slotsInWeek(week: Long): List<Int> = slots.filter { it.validIn(week) }.map { it.slot }.distinct().sorted()

        override fun toString(): String = "ScheduleSummary(seq=$seq, network=$network, weeks=$firstWeek..$lastWeek)"
    }

    /** A frozen layout: the digest stored at purchase time and the number of positions. */
    class Layout(digest: ByteArray, val positions: Int) {
        private val digestBytes = digest.copyOf()
        fun digest(): ByteArray = digestBytes.copyOf()
        override fun toString(): String = "Layout(positions=$positions)"
    }

    /** A token that verified offline: its (kind, epoch, slot) and nullifier (a secret, R10). */
    class VerifiedToken(val kind: Int, val epoch: Long, val slot: Int?, nullifier: ByteArray) {
        private val nullifierBytes = nullifier.copyOf()
        fun nullifier(): ByteArray = nullifierBytes.copyOf()
        override fun toString(): String = "VerifiedToken(kind=$kind, epoch=$epoch)"
    }

    data class AddressInfo(val network: Int, val type: Int)

    /** Strict big-endian reader: any read past the end is `malformed_response`. */
    private class Reader(private val raw: ByteArray) {
        var at = 0
            private set

        fun u8(): Int {
            if (at + 1 > raw.size) throw NetworkException(MALFORMED)
            return raw[at++].toInt() and 0xff
        }

        fun u16(): Int = (u8() shl 8) or u8()

        /** An unsigned 64-bit value that must fit a non-negative Long. */
        fun u64(): Long {
            var v = 0L
            repeat(8) { v = (v shl 8) or u8().toLong() }
            if (v < 0) throw NetworkException(MALFORMED)
            return v
        }

        fun bytes(n: Int): ByteArray {
            if (n < 0 || at + n > raw.size) throw NetworkException(MALFORMED)
            return raw.copyOfRange(at, at + n).also { at += n }
        }

        fun end() {
            if (at != raw.size) throw NetworkException(MALFORMED)
        }
    }

    private fun kindOk(kind: Int) = kind in KIND_ACCESS..KIND_CREDIT

    /** Layout of `client-core/net/src/entitlement.rs` `schedule_summary`. */
    internal fun decodeSummary(raw: ByteArray): ScheduleSummary {
        val r = Reader(raw)
        val digest = r.bytes(32)
        val seq = r.u64()
        val network = r.u8()
        val first = r.u64()
        val last = r.u64()
        if (seq < 1 || network !in 1..3 || last < first) throw NetworkException(MALFORMED)
        val constants = Constants(
            confirmations = r.u8(),
            invoiceBlocks = r.u16(),
            graceBlocks = r.u16(),
            accessPerSlot = r.u8(),
            trialPerSlot = r.u8(),
            invitesPerPack = r.u8(),
            creditsPerFreePack = r.u8(),
            minClaimCredits = r.u8(),
            maxClaimCredits = r.u8(),
            earlyWindowHours = r.u8(),
            capabilityQuotaBytes = r.u64(),
        )
        val slots = List(r.u8()) {
            val slot = r.u8()
            val from = r.u64()
            val until = r.u64()
            val text = String(r.bytes(r.u8()), Charsets.US_ASCII)
            if (slot > MAX_SLOT || (until != 0L && until <= from)) throw NetworkException(MALFORMED)
            val onion = try {
                OnionAddress.parse(text)
            } catch (e: IllegalArgumentException) {
                throw NetworkException(MALFORMED)
            }
            Slot(slot, from, until, onion)
        }
        val prices = List(r.u16()) { Price(r.u64(), r.u64()) }
        val keys = List(r.u16()) {
            val kind = r.u8()
            val epoch = r.u64()
            if (!kindOk(kind)) throw NetworkException(MALFORMED)
            KeyId(kind, epoch, r.bytes(32))
        }
        val revoked = List(r.u16()) {
            val kind = r.u8()
            if (!kindOk(kind)) throw NetworkException(MALFORMED)
            Revoked(kind, r.u64())
        }
        r.end()
        return ScheduleSummary(digest, seq, network, first, last, constants, slots, prices, keys, revoked)
    }

    /** `digest(32) || N(4)`, N >= 1. */
    internal fun decodeLayout(raw: ByteArray): Layout {
        val r = Reader(raw)
        val digest = r.bytes(32)
        val n = (r.u16() shl 16) or r.u16()
        r.end()
        if (n < 1) throw NetworkException(MALFORMED)
        return Layout(digest, n)
    }

    /** `kind(1) || epoch(8) || slot(1) || nullifier(32)` for a token checked as [kind]. */
    internal fun decodeVerified(raw: ByteArray, kind: Int): VerifiedToken {
        val r = Reader(raw)
        val gotKind = r.u8()
        val epoch = r.u64()
        val slot = r.u8()
        val nullifier = r.bytes(32)
        r.end()
        if (gotKind != kind) throw NetworkException(MALFORMED)
        val slotOk = if (kind == KIND_ACCESS) slot <= MAX_SLOT else slot == NO_SLOT
        if (!slotOk) throw NetworkException(MALFORMED)
        return VerifiedToken(kind, epoch, if (kind == KIND_ACCESS) slot else null, nullifier)
    }

    /** `(network << 8) | type`, network 1..3, type standard or subaddress. */
    internal fun decodeAddress(code: Int): AddressInfo {
        val network = code ushr 8
        val type = code and 0xff
        if (network !in 1..3 || (type != ADDRESS_STANDARD && type != ADDRESS_SUBADDRESS)) throw NetworkException(MALFORMED)
        return AddressInfo(network, type)
    }

    @JvmStatic private external fun nativeScheduleSummary(): ByteArray
    @JvmStatic private external fun nativeLayoutDigest(product: Int, index: Long): ByteArray
    @JvmStatic private external fun nativeVerifyToken(token: ByteArray, kind: Int): ByteArray
    @JvmStatic private external fun nativeValidateAddress(address: String, purpose: Int): Int
    @JvmStatic private external fun nativePaymentUri(subaddress: String, amountAtomic: Long): String
}
